Expose workspace readLock to python to prevent segfaults

joseph-torsney opened the pull request:
#31367 finds a crash that occurs when running `LoadInstrument` multiple times while the `SliceViewer` is open. This is because, on the first run of `LoadInstrument`, the workspace it is running on changes. The `SliceViewerADSObserver` detects this and begins to draw the view, where it accesses some spectra information on the workspace. While this update has been happening, `LoadInstrument` modifies the same workspace again which leads to a segfault.

This PR exposes workspace `readLock` and `unlock` to the python API so that widgets can lock the workspace while they update themselves.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**
1. Load some data, e.g. `INTER38417`
2. Open sliceviewer on this workspace
3. Run an algorithm repeatedly that modifies this workspace e.g. by using the following script
```
ws = mtd["INTER38417"]
while True:
    LoadInstrument(Workspace=ws, RewriteSpectraMap=False, InstrumentName="INTER")
``` 
4. Workbench shouldn't crash

Fixes #31367

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

joseph-torsney committed:
Expose workspace readLock to python re #31367

martyngigg commented on Framework/Kernel/inc/MantidKernel/DataItem.h:
```diff
@@ -54,6 +54,11 @@ class MANTID_KERNEL_DLL DataItem {
 
+  /// Acquires a read lock. If another thread currently holds a write lock,
+  /// waits until the write lock is released.
+  void readLock();
```
Can we inline this instead as it's just a pass through, i.e.

```cpp
inline void readLock() { getLock()->readLock(); }
```

joseph-torsney replied:
When I do this inline I get `Pointer to incomplete class type is not allowed` on the `getLock()` call. It resolves if I do `include <Poco/RWLock.h>` in this header, but doesn't that mean everything that includes `DataItem.h` will also include this bit of Poco?

martyngigg replied:
Ah yes that's a fair point. Maybe it is better to keep this more encapsulated. Let's leave it as you have it 👍🏻

martyngigg commented on qt/python/mantidqt/utils/qt/qappthreadcall.py:
```diff
@@ -47,7 +48,10 @@ def __call__(self, *args, **kwargs):
             self._ensure_self_on_qapp_thread()
             self._store_function_args(args, kwargs)
-            QMetaObject.invokeMethod(self, "on_call", Qt.BlockingQueuedConnection)
+            if self.blocking:
```
It can be good practice to avoid repetition of the function/method call itself in cases like this and instead just adjust the parameter so that when reading the code it's obvious that only a parameter differs. In this case it's not too bad but for more parameters you need to more carefully read both clauses to see if the other arguments are the same, e.g.
```python
            connection_type = Qt.BlockingQueuedConnection if self.blocking else Qt.AutoConnection
            QMetaObject.invokeMethod(self, "on_call", connection_type)
```

martyngigg commented on qt/python/mantidqt/widgets/sliceviewer/presenter.py:
```diff
@@ -396,10 +396,13 @@ def refresh_view(self):
+        ws = self.model._get_ws()
         # we don't want to use model.get_ws for the image info widget as this needs
         # extra arguments depending on workspace type.
-        self.view.data_view.image_info_widget.setWorkspace(self.model._get_ws())
```
This is a really nice fix. Just one thing to be careful about when dealing with locks - exceptions. In you Python export methods we use the `ReleaseGlobalInterpreterLock` struct to ensure we always lock the GIL again.
In this case if anything throws an exception after the lock then the unlock won't run. In Python we can use `try...finally` to do something similar to RAII so that the `finally` clause always runs regardless if there is an exception, e.g.
 
```python
ws.readLock()
try:
    self.view.data_view.image_info_widget.setWorkspace(ws)
    self.new_plot()
finally:
    ws.unlock()
```

martyngigg reviewed (dismissed):
A neat fix to our race condition problem. Just a couple of suggestions for improving the robustness of the code.

joseph-torsney committed:
Tidy up code after review re #31365

martyngigg commented:
Thanks for making those changes. Could you add a snippet of test instructions?

joseph-torsney commented:
Ah my bad, forgot them while taking it out of draft. Thanks for the review and some good tips!

martyngigg commented:
It looks like we might still have an issue here. In debug a few iterations of `LoadInstrument` now get through but in release it falls over after first one for me.

joseph-torsney commented:
Should be fixed now. When `refresh_view` was called by the ADSObserver, it does a `readLock()`. This solved most of the crashes, but the view update is split between this and [`canvas.draw_idle()`](https://github.com/mantidproject/mantid/blob/f9efc92eba5645c0e64484504df53953a485cba0/qt/python/mantidqt/widgets/sliceviewer/view.py#L360). `draw_idle` puts the draw on the qt event loop. Then `refresh_view` unlocks the workspace. While we're waiting for the event loop to get around to the draw, another call to `LoadInstrument` sneaks in and modifies the workspace again, which caused another segfault. 

The solution is to replace `draw_idle` with `draw`, so that it is called immediately within the original `readLock`, and the entire update happens at the same time.

joseph-torsney committed:
Sliceviewer remove delayed_refresh re #31367

Github pull request 30242 introduces delayed refresh. This is no longer
necessary with the addition of non-blocking QAppThreadCalls. See gitub
pull request 32204 for more details.

martyngigg committed:
Use .draw and not .draw_idle to redraw

Guarantees a sequential flow

DanielMurphy22 reviewed (approved):
This seems to work in release and debug! Good job!
